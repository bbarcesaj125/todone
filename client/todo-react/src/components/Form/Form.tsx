import React from "react";
import { useState } from "react";
import styled from "styled-components";

interface Props {
  addItem: AddItem;
}

interface LabelProps {
  for: string;
}

// Styles
const FormElement = styled.form`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

const Heading = styled.h2``;

const Label = styled.label<LabelProps>``;

const Input = styled.input`
  width: 100%;
  height: 50px;
  background-color: ${(props) => props.theme.switchBackground};
  color: ${(props) => props.theme.text};
  border: 1px solid ${(props) => props.theme.border};
  border-radius: 5px;
`;

const TextArea = styled.textarea`
  width: 100%;
  height: 100px;
  background-color: ${(props) => props.theme.switchBackground};
  color: ${(props) => props.theme.text};
  border: 1px solid ${(props) => props.theme.border};
  border-radius: 5px;
`;

const Button = styled.button`
  width: 100%;
  height: 50px;
  margin: 10px;
  background-color: ${(props) => props.theme.switchBackground};
  color: ${(props) => props.theme.text};
  border: 1px solid ${(props) => props.theme.border};
  border-radius: 5px;
  cursor: pointer;
`;

const Form: React.FC<Props> = ({ addItem }) => {
  const initialValues = {
    title: "",
    description: "",
  };
  const [inputs, setInputs] = useState(initialValues);

  const handleChange = (
    event:
      | React.ChangeEvent<HTMLInputElement>
      | React.ChangeEvent<HTMLTextAreaElement>
  ) => {
    const name = event.target.name;
    const value = event.target.value;
    setInputs((values) => ({ ...values, [name]: value }));
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    addItem(inputs);
    // Form reset
    setInputs(initialValues);
  };
  return (
    <FormElement onSubmit={handleSubmit}>
      <Heading>
        <Label for="title">Please enter the task&apos;s title:</Label>
      </Heading>

      <Input
        id="title"
        type="text"
        name="title"
        value={inputs.title}
        onChange={handleChange}
        required
        placeholder="Title"
        aria-label="title"
      />
      <Heading>
        <Label for="description">
          Please enter the task&apos;s description (optional):
        </Label>
      </Heading>
      <TextArea
        id="description"
        name="description"
        value={inputs.description}
        onChange={handleChange}
        placeholder="Description"
        aria-label="description"
      ></TextArea>

      <Button type="submit">Add</Button>
    </FormElement>
  );
};

export default Form;
