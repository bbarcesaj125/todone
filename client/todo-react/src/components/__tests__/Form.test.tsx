import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import Form from "../Form";

const mockedAddItem = jest.fn();

describe("Form", () => {
  const { asFragment } = render(<Form addItem={mockedAddItem}></Form>);
  it("Form responds correctly to events", () => {
    const button = screen.getByRole("button");
    const input = screen.getByRole("textbox", { name: /title/i });
    fireEvent.click(button);
    fireEvent.change(input, { target: { value: "Title test" } });
    expect(mockedAddItem).toBeCalledTimes(1);

    const titleInput = screen.getByDisplayValue("Title test");
    expect(titleInput).toBeInTheDocument;
  });

  it("Form renders correctly", () => {
    expect(asFragment()).toMatchSnapshot();
  });
});
