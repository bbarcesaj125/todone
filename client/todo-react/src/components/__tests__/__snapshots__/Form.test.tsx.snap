// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Form renders correctly 1`] = `<DocumentFragment />`;

exports[`SingleView should display information about a single task 1`] = `
<DocumentFragment>
  <form>
    <h2>
      <label>
        Please enter the task's title:
      </label>
    </h2>
    <input
      aria-label="title"
      name="title"
      placeholder="Title"
      required=""
      type="text"
      value=""
    />
    <h2>
      <label>
        Please enter the task's description (optional):
      </label>
    </h2>
    <textarea
      aria-label="description"
      name="description"
      placeholder="Description"
    />
    <button
      type="submit"
    >
      Add
    </button>
  </form>
</DocumentFragment>
`;
